// google-ads-api/build/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Customer = exports.ResourceNames = exports.parse = exports.toMicros = exports.fromMicros = exports.protobuf = exports.longrunning = exports.services = exports.resources = exports.errors = exports.fields = exports.enums = exports.common = exports.GoogleAdsApi = void 0;
// Core library client
var client_1 = require("./client");
Object.defineProperty(exports, "GoogleAdsApi", { enumerable: true, get: function () { return client_1.Client; } });
// Compiled proto types
var index_1 = require("./protos/index");
Object.defineProperty(exports, "common", { enumerable: true, get: function () { return index_1.common; } });
Object.defineProperty(exports, "enums", { enumerable: true, get: function () { return index_1.enums; } });
Object.defineProperty(exports, "fields", { enumerable: true, get: function () { return index_1.fields; } });
Object.defineProperty(exports, "errors", { enumerable: true, get: function () { return index_1.errors; } });
Object.defineProperty(exports, "resources", { enumerable: true, get: function () { return index_1.resources; } });
Object.defineProperty(exports, "services", { enumerable: true, get: function () { return index_1.services; } });
Object.defineProperty(exports, "longrunning", { enumerable: true, get: function () { return index_1.longrunning; } });
Object.defineProperty(exports, "protobuf", { enumerable: true, get: function () { return index_1.protobuf; } });
// Util functions
var utils_1 = require("./utils");
Object.defineProperty(exports, "fromMicros", { enumerable: true, get: function () { return utils_1.fromMicros; } });
Object.defineProperty(exports, "toMicros", { enumerable: true, get: function () { return utils_1.toMicros; } });
var parser_1 = require("./parser");
Object.defineProperty(exports, "parse", { enumerable: true, get: function () { return parser_1.parse; } });
exports.ResourceNames = __importStar(require("./protos/autogen/resourceNames"));
var customer_1 = require("./customer");
Object.defineProperty(exports, "Customer", { enumerable: true, get: function () { return customer_1.Customer; } });
