// google-ads-api/build/src/protos/autogen/resourceNames.js
"use strict";
/* Autogenerated File! Do Not Edit */
Object.defineProperty(exports, "__esModule", { value: true });
exports.accessibleBiddingStrategy = accessibleBiddingStrategy;
exports.accountBudget = accountBudget;
exports.accountBudgetProposal = accountBudgetProposal;
exports.accountLink = accountLink;
exports.ad = ad;
exports.adGroup = adGroup;
exports.adGroupAd = adGroupAd;
exports.adGroupAdAssetCombinationView = adGroupAdAssetCombinationView;
exports.adGroupAdAssetView = adGroupAdAssetView;
exports.adGroupAdLabel = adGroupAdLabel;
exports.adGroupAsset = adGroupAsset;
exports.adGroupAssetSet = adGroupAssetSet;
exports.adGroupAudienceView = adGroupAudienceView;
exports.adGroupBidModifier = adGroupBidModifier;
exports.adGroupCriterion = adGroupCriterion;
exports.adGroupCriterionCustomizer = adGroupCriterionCustomizer;
exports.adGroupCriterionLabel = adGroupCriterionLabel;
exports.adGroupCriterionSimulation = adGroupCriterionSimulation;
exports.adGroupCustomizer = adGroupCustomizer;
exports.adGroupLabel = adGroupLabel;
exports.adGroupSimulation = adGroupSimulation;
exports.adParameter = adParameter;
exports.adScheduleView = adScheduleView;
exports.ageRangeView = ageRangeView;
exports.aiMaxSearchTermAdCombinationView = aiMaxSearchTermAdCombinationView;
exports.androidPrivacySharedKeyGoogleAdGroup = androidPrivacySharedKeyGoogleAdGroup;
exports.androidPrivacySharedKeyGoogleCampaign = androidPrivacySharedKeyGoogleCampaign;
exports.androidPrivacySharedKeyGoogleNetworkType = androidPrivacySharedKeyGoogleNetworkType;
exports.asset = asset;
exports.assetFieldTypeView = assetFieldTypeView;
exports.assetGroup = assetGroup;
exports.assetGroupAsset = assetGroupAsset;
exports.assetGroupListingGroupFilter = assetGroupListingGroupFilter;
exports.assetGroupProductGroupView = assetGroupProductGroupView;
exports.assetGroupSignal = assetGroupSignal;
exports.assetGroupTopCombinationView = assetGroupTopCombinationView;
exports.assetSet = assetSet;
exports.assetSetAsset = assetSetAsset;
exports.assetSetTypeView = assetSetTypeView;
exports.audience = audience;
exports.batchJob = batchJob;
exports.biddingDataExclusion = biddingDataExclusion;
exports.biddingSeasonalityAdjustment = biddingSeasonalityAdjustment;
exports.biddingStrategy = biddingStrategy;
exports.biddingStrategySimulation = biddingStrategySimulation;
exports.billingSetup = billingSetup;
exports.callView = callView;
exports.campaign = campaign;
exports.campaignAggregateAssetView = campaignAggregateAssetView;
exports.campaignAsset = campaignAsset;
exports.campaignAssetSet = campaignAssetSet;
exports.campaignAudienceView = campaignAudienceView;
exports.campaignBidModifier = campaignBidModifier;
exports.campaignBudget = campaignBudget;
exports.campaignConversionGoal = campaignConversionGoal;
exports.campaignCriterion = campaignCriterion;
exports.campaignCustomizer = campaignCustomizer;
exports.campaignDraft = campaignDraft;
exports.campaignGroup = campaignGroup;
exports.campaignLabel = campaignLabel;
exports.campaignLifecycleGoal = campaignLifecycleGoal;
exports.campaignSearchTermInsight = campaignSearchTermInsight;
exports.campaignSearchTermView = campaignSearchTermView;
exports.campaignSharedSet = campaignSharedSet;
exports.campaignSimulation = campaignSimulation;
exports.carrierConstant = carrierConstant;
exports.changeEvent = changeEvent;
exports.changeStatus = changeStatus;
exports.channelAggregateAssetView = channelAggregateAssetView;
exports.clickView = clickView;
exports.combinedAudience = combinedAudience;
exports.contentCriterionView = contentCriterionView;
exports.conversionAction = conversionAction;
exports.conversionCustomVariable = conversionCustomVariable;
exports.conversionGoalCampaignConfig = conversionGoalCampaignConfig;
exports.conversionValueRule = conversionValueRule;
exports.conversionValueRuleSet = conversionValueRuleSet;
exports.currencyConstant = currencyConstant;
exports.customAudience = customAudience;
exports.customConversionGoal = customConversionGoal;
exports.customInterest = customInterest;
exports.customer = customer;
exports.customerAsset = customerAsset;
exports.customerAssetSet = customerAssetSet;
exports.customerClient = customerClient;
exports.customerClientLink = customerClientLink;
exports.customerConversionGoal = customerConversionGoal;
exports.customerCustomizer = customerCustomizer;
exports.customerLabel = customerLabel;
exports.customerLifecycleGoal = customerLifecycleGoal;
exports.customerManagerLink = customerManagerLink;
exports.customerNegativeCriterion = customerNegativeCriterion;
exports.customerSearchTermInsight = customerSearchTermInsight;
exports.customerSkAdNetworkConversionValueSchema = customerSkAdNetworkConversionValueSchema;
exports.customerUserAccess = customerUserAccess;
exports.customerUserAccessInvitation = customerUserAccessInvitation;
exports.customizerAttribute = customizerAttribute;
exports.dataLink = dataLink;
exports.detailContentSuitabilityPlacementView = detailContentSuitabilityPlacementView;
exports.detailPlacementView = detailPlacementView;
exports.detailedDemographic = detailedDemographic;
exports.displayKeywordView = displayKeywordView;
exports.distanceView = distanceView;
exports.domainCategory = domainCategory;
exports.dynamicSearchAdsSearchTermView = dynamicSearchAdsSearchTermView;
exports.expandedLandingPageView = expandedLandingPageView;
exports.experiment = experiment;
exports.experimentArm = experimentArm;
exports.finalUrlExpansionAssetView = finalUrlExpansionAssetView;
exports.genderView = genderView;
exports.geoTargetConstant = geoTargetConstant;
exports.geographicView = geographicView;
exports.googleAdsField = googleAdsField;
exports.groupContentSuitabilityPlacementView = groupContentSuitabilityPlacementView;
exports.groupPlacementView = groupPlacementView;
exports.hotelGroupView = hotelGroupView;
exports.hotelPerformanceView = hotelPerformanceView;
exports.hotelReconciliation = hotelReconciliation;
exports.incomeRangeView = incomeRangeView;
exports.invoice = invoice;
exports.keywordPlan = keywordPlan;
exports.keywordPlanAdGroup = keywordPlanAdGroup;
exports.keywordPlanAdGroupKeyword = keywordPlanAdGroupKeyword;
exports.keywordPlanCampaign = keywordPlanCampaign;
exports.keywordPlanCampaignKeyword = keywordPlanCampaignKeyword;
exports.keywordThemeConstant = keywordThemeConstant;
exports.keywordView = keywordView;
exports.label = label;
exports.landingPageView = landingPageView;
exports.languageConstant = languageConstant;
exports.leadFormSubmissionData = leadFormSubmissionData;
exports.lifeEvent = lifeEvent;
exports.localServicesEmployee = localServicesEmployee;
exports.localServicesLead = localServicesLead;
exports.localServicesLeadConversation = localServicesLeadConversation;
exports.localServicesVerificationArtifact = localServicesVerificationArtifact;
exports.locationInterestView = locationInterestView;
exports.locationView = locationView;
exports.managedPlacementView = managedPlacementView;
exports.mediaFile = mediaFile;
exports.mobileAppCategoryConstant = mobileAppCategoryConstant;
exports.mobileDeviceConstant = mobileDeviceConstant;
exports.offlineConversionUploadClientSummary = offlineConversionUploadClientSummary;
exports.offlineConversionUploadConversionActionSummary = offlineConversionUploadConversionActionSummary;
exports.offlineUserDataJob = offlineUserDataJob;
exports.operatingSystemVersionConstant = operatingSystemVersionConstant;
exports.paidOrganicSearchTermView = paidOrganicSearchTermView;
exports.parentalStatusView = parentalStatusView;
exports.paymentsAccount = paymentsAccount;
exports.perStoreView = perStoreView;
exports.performanceMaxPlacementView = performanceMaxPlacementView;
exports.productCategoryConstant = productCategoryConstant;
exports.productGroupView = productGroupView;
exports.productLink = productLink;
exports.productLinkInvitation = productLinkInvitation;
exports.qualifyingQuestion = qualifyingQuestion;
exports.recommendation = recommendation;
exports.recommendationSubscription = recommendationSubscription;
exports.remarketingAction = remarketingAction;
exports.searchTermView = searchTermView;
exports.sharedCriterion = sharedCriterion;
exports.sharedSet = sharedSet;
exports.shoppingPerformanceView = shoppingPerformanceView;
exports.shoppingProduct = shoppingProduct;
exports.smartCampaignSearchTermView = smartCampaignSearchTermView;
exports.smartCampaignSetting = smartCampaignSetting;
exports.thirdPartyAppAnalyticsLink = thirdPartyAppAnalyticsLink;
exports.topicConstant = topicConstant;
exports.topicView = topicView;
exports.travelActivityGroupView = travelActivityGroupView;
exports.travelActivityPerformanceView = travelActivityPerformanceView;
exports.userInterest = userInterest;
exports.userList = userList;
exports.userListCustomerType = userListCustomerType;
exports.userLocationView = userLocationView;
exports.video = video;
exports.webpageView = webpageView;
/**
 * @param {string | number} customerId
 * @param {string | number} biddingStrategyId
 * @returns `AccessibleBiddingStrategyResourceName`
 * @example const accessibleBiddingStrategy: ResourceNames.AccessibleBiddingStrategyResourceName = ResourceNames.accessibleBiddingStrategy(10987417, 21974834)
 */
function accessibleBiddingStrategy(customerId, biddingStrategyId) {
    return `customers/${customerId}/accessibleBiddingStrategies/${biddingStrategyId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} accountBudgetId
 * @returns `AccountBudgetResourceName`
 * @example const accountBudget: ResourceNames.AccountBudgetResourceName = ResourceNames.accountBudget(10987417, 21974834)
 */
function accountBudget(customerId, accountBudgetId) {
    return `customers/${customerId}/accountBudgets/${accountBudgetId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} accountBudgetProposalId
 * @returns `AccountBudgetProposalResourceName`
 * @example const accountBudgetProposal: ResourceNames.AccountBudgetProposalResourceName = ResourceNames.accountBudgetProposal(10987417, 21974834)
 */
function accountBudgetProposal(customerId, accountBudgetProposalId) {
    return `customers/${customerId}/accountBudgetProposals/${accountBudgetProposalId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} accountLinkId
 * @returns `AccountLinkResourceName`
 * @example const accountLink: ResourceNames.AccountLinkResourceName = ResourceNames.accountLink(10987417, 21974834)
 */
function accountLink(customerId, accountLinkId) {
    return `customers/${customerId}/accountLinks/${accountLinkId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adId
 * @returns `AdResourceName`
 * @example const ad: ResourceNames.AdResourceName = ResourceNames.ad(10987417, 21974834)
 */
function ad(customerId, adId) {
    return `customers/${customerId}/ads/${adId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @returns `AdGroupResourceName`
 * @example const adGroup: ResourceNames.AdGroupResourceName = ResourceNames.adGroup(10987417, 21974834)
 */
function adGroup(customerId, adGroupId) {
    return `customers/${customerId}/adGroups/${adGroupId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} adId
 * @returns `AdGroupAdResourceName`
 * @example const adGroupAd: ResourceNames.AdGroupAdResourceName = ResourceNames.adGroupAd(10987417, 21974834, 43949668)
 */
function adGroupAd(customerId, adGroupId, adId) {
    return `customers/${customerId}/adGroupAds/${adGroupId}~${adId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} adId
 * @param {string | number} assetCombinationIdLow
 * @param {string | number} assetCombinationIdHigh
 * @returns `AdGroupAdAssetCombinationViewResourceName`
 * @example const adGroupAdAssetCombinationView: ResourceNames.AdGroupAdAssetCombinationViewResourceName = ResourceNames.adGroupAdAssetCombinationView(10987417, 21974834, 43949668, 87899336, 175798672)
 */
function adGroupAdAssetCombinationView(customerId, adGroupId, adId, assetCombinationIdLow, assetCombinationIdHigh) {
    return `customers/${customerId}/adGroupAdAssetCombinationViews/${adGroupId}~${adId}~${assetCombinationIdLow}~${assetCombinationIdHigh}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} adId
 * @param {string | number} assetId
 * @param {string | number} fieldType
 * @returns `AdGroupAdAssetViewResourceName`
 * @example const adGroupAdAssetView: ResourceNames.AdGroupAdAssetViewResourceName = ResourceNames.adGroupAdAssetView(10987417, 21974834, 43949668, 87899336, 175798672)
 */
function adGroupAdAssetView(customerId, adGroupId, adId, assetId, fieldType) {
    return `customers/${customerId}/adGroupAdAssetViews/${adGroupId}~${adId}~${assetId}~${fieldType}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} adId
 * @param {string | number} labelId
 * @returns `AdGroupAdLabelResourceName`
 * @example const adGroupAdLabel: ResourceNames.AdGroupAdLabelResourceName = ResourceNames.adGroupAdLabel(10987417, 21974834, 43949668, 87899336)
 */
function adGroupAdLabel(customerId, adGroupId, adId, labelId) {
    return `customers/${customerId}/adGroupAdLabels/${adGroupId}~${adId}~${labelId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} assetId
 * @param {string | number} fieldType
 * @returns `AdGroupAssetResourceName`
 * @example const adGroupAsset: ResourceNames.AdGroupAssetResourceName = ResourceNames.adGroupAsset(10987417, 21974834, 43949668, 87899336)
 */
function adGroupAsset(customerId, adGroupId, assetId, fieldType) {
    return `customers/${customerId}/adGroupAssets/${adGroupId}~${assetId}~${fieldType}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} assetSetId
 * @returns `AdGroupAssetSetResourceName`
 * @example const adGroupAssetSet: ResourceNames.AdGroupAssetSetResourceName = ResourceNames.adGroupAssetSet(10987417, 21974834, 43949668)
 */
function adGroupAssetSet(customerId, adGroupId, assetSetId) {
    return `customers/${customerId}/adGroupAssetSets/${adGroupId}~${assetSetId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @returns `AdGroupAudienceViewResourceName`
 * @example const adGroupAudienceView: ResourceNames.AdGroupAudienceViewResourceName = ResourceNames.adGroupAudienceView(10987417, 21974834, 43949668)
 */
function adGroupAudienceView(customerId, adGroupId, criterionId) {
    return `customers/${customerId}/adGroupAudienceViews/${adGroupId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @returns `AdGroupBidModifierResourceName`
 * @example const adGroupBidModifier: ResourceNames.AdGroupBidModifierResourceName = ResourceNames.adGroupBidModifier(10987417, 21974834, 43949668)
 */
function adGroupBidModifier(customerId, adGroupId, criterionId) {
    return `customers/${customerId}/adGroupBidModifiers/${adGroupId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @returns `AdGroupCriterionResourceName`
 * @example const adGroupCriterion: ResourceNames.AdGroupCriterionResourceName = ResourceNames.adGroupCriterion(10987417, 21974834, 43949668)
 */
function adGroupCriterion(customerId, adGroupId, criterionId) {
    return `customers/${customerId}/adGroupCriteria/${adGroupId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @param {string | number} customizerAttributeId
 * @returns `AdGroupCriterionCustomizerResourceName`
 * @example const adGroupCriterionCustomizer: ResourceNames.AdGroupCriterionCustomizerResourceName = ResourceNames.adGroupCriterionCustomizer(10987417, 21974834, 43949668, 87899336)
 */
function adGroupCriterionCustomizer(customerId, adGroupId, criterionId, customizerAttributeId) {
    return `customers/${customerId}/adGroupCriterionCustomizers/${adGroupId}~${criterionId}~${customizerAttributeId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @param {string | number} labelId
 * @returns `AdGroupCriterionLabelResourceName`
 * @example const adGroupCriterionLabel: ResourceNames.AdGroupCriterionLabelResourceName = ResourceNames.adGroupCriterionLabel(10987417, 21974834, 43949668, 87899336)
 */
function adGroupCriterionLabel(customerId, adGroupId, criterionId, labelId) {
    return `customers/${customerId}/adGroupCriterionLabels/${adGroupId}~${criterionId}~${labelId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @param {string | number} type
 * @param {string | number} modificationMethod
 * @param {string | number} startDate
 * @param {string | number} endDate
 * @returns `AdGroupCriterionSimulationResourceName`
 * @example const adGroupCriterionSimulation: ResourceNames.AdGroupCriterionSimulationResourceName = ResourceNames.adGroupCriterionSimulation(10987417, 21974834, 43949668, 87899336, 175798672, 351597344, 703194688)
 */
function adGroupCriterionSimulation(customerId, adGroupId, criterionId, type, modificationMethod, startDate, endDate) {
    return `customers/${customerId}/adGroupCriterionSimulations/${adGroupId}~${criterionId}~${type}~${modificationMethod}~${startDate}~${endDate}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} customizerAttributeId
 * @returns `AdGroupCustomizerResourceName`
 * @example const adGroupCustomizer: ResourceNames.AdGroupCustomizerResourceName = ResourceNames.adGroupCustomizer(10987417, 21974834, 43949668)
 */
function adGroupCustomizer(customerId, adGroupId, customizerAttributeId) {
    return `customers/${customerId}/adGroupCustomizers/${adGroupId}~${customizerAttributeId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} labelId
 * @returns `AdGroupLabelResourceName`
 * @example const adGroupLabel: ResourceNames.AdGroupLabelResourceName = ResourceNames.adGroupLabel(10987417, 21974834, 43949668)
 */
function adGroupLabel(customerId, adGroupId, labelId) {
    return `customers/${customerId}/adGroupLabels/${adGroupId}~${labelId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} type
 * @param {string | number} modificationMethod
 * @param {string | number} startDate
 * @param {string | number} endDate
 * @returns `AdGroupSimulationResourceName`
 * @example const adGroupSimulation: ResourceNames.AdGroupSimulationResourceName = ResourceNames.adGroupSimulation(10987417, 21974834, 43949668, 87899336, 175798672, 351597344)
 */
function adGroupSimulation(customerId, adGroupId, type, modificationMethod, startDate, endDate) {
    return `customers/${customerId}/adGroupSimulations/${adGroupId}~${type}~${modificationMethod}~${startDate}~${endDate}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @param {string | number} parameterIndex
 * @returns `AdParameterResourceName`
 * @example const adParameter: ResourceNames.AdParameterResourceName = ResourceNames.adParameter(10987417, 21974834, 43949668, 87899336)
 */
function adParameter(customerId, adGroupId, criterionId, parameterIndex) {
    return `customers/${customerId}/adParameters/${adGroupId}~${criterionId}~${parameterIndex}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} criterionId
 * @returns `AdScheduleViewResourceName`
 * @example const adScheduleView: ResourceNames.AdScheduleViewResourceName = ResourceNames.adScheduleView(10987417, 21974834, 43949668)
 */
function adScheduleView(customerId, campaignId, criterionId) {
    return `customers/${customerId}/adScheduleViews/${campaignId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @returns `AgeRangeViewResourceName`
 * @example const ageRangeView: ResourceNames.AgeRangeViewResourceName = ResourceNames.ageRangeView(10987417, 21974834, 43949668)
 */
function ageRangeView(customerId, adGroupId, criterionId) {
    return `customers/${customerId}/ageRangeViews/${adGroupId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} searchTerm
 * @param {string | number} landingPage
 * @param {string | number} headline
 * @returns `AiMaxSearchTermAdCombinationViewResourceName`
 * @example const aiMaxSearchTermAdCombinationView: ResourceNames.AiMaxSearchTermAdCombinationViewResourceName = ResourceNames.aiMaxSearchTermAdCombinationView(10987417, 21974834, 43949668, 87899336, 175798672)
 */
function aiMaxSearchTermAdCombinationView(customerId, adGroupId, searchTerm, landingPage, headline) {
    return `customers/${customerId}/aiMaxSearchTermAdCombinationViews/${adGroupId}~${searchTerm}~${landingPage}~${headline}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} adGroupId
 * @param {string | number} androidPrivacyInteractionType
 * @param {string | number} androidPrivacyNetworkType
 * @param {string | number} androidPrivacyInteractionDate
 * @returns `AndroidPrivacySharedKeyGoogleAdGroupResourceName`
 * @example const androidPrivacySharedKeyGoogleAdGroup: ResourceNames.AndroidPrivacySharedKeyGoogleAdGroupResourceName = ResourceNames.androidPrivacySharedKeyGoogleAdGroup(10987417, 21974834, 43949668, 87899336, 175798672, 351597344)
 */
function androidPrivacySharedKeyGoogleAdGroup(customerId, campaignId, adGroupId, androidPrivacyInteractionType, androidPrivacyNetworkType, androidPrivacyInteractionDate) {
    return `customers/${customerId}/androidPrivacySharedKeyGoogleAdGroups/${campaignId}~${adGroupId}~${androidPrivacyInteractionType}~${androidPrivacyNetworkType}~${androidPrivacyInteractionDate}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} androidPrivacyInteractionType
 * @param {string | number} androidPrivacyInteractionDate
 * @returns `AndroidPrivacySharedKeyGoogleCampaignResourceName`
 * @example const androidPrivacySharedKeyGoogleCampaign: ResourceNames.AndroidPrivacySharedKeyGoogleCampaignResourceName = ResourceNames.androidPrivacySharedKeyGoogleCampaign(10987417, 21974834, 43949668, 87899336)
 */
function androidPrivacySharedKeyGoogleCampaign(customerId, campaignId, androidPrivacyInteractionType, androidPrivacyInteractionDate) {
    return `customers/${customerId}/androidPrivacySharedKeyGoogleCampaigns/${campaignId}~${androidPrivacyInteractionType}~${androidPrivacyInteractionDate}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} androidPrivacyInteractionType
 * @param {string | number} androidPrivacyNetworkType
 * @param {string | number} androidPrivacyInteractionDate
 * @returns `AndroidPrivacySharedKeyGoogleNetworkTypeResourceName`
 * @example const androidPrivacySharedKeyGoogleNetworkType: ResourceNames.AndroidPrivacySharedKeyGoogleNetworkTypeResourceName = ResourceNames.androidPrivacySharedKeyGoogleNetworkType(10987417, 21974834, 43949668, 87899336, 175798672)
 */
function androidPrivacySharedKeyGoogleNetworkType(customerId, campaignId, androidPrivacyInteractionType, androidPrivacyNetworkType, androidPrivacyInteractionDate) {
    return `customers/${customerId}/androidPrivacySharedKeyGoogleNetworkTypes/${campaignId}~${androidPrivacyInteractionType}~${androidPrivacyNetworkType}~${androidPrivacyInteractionDate}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} assetId
 * @returns `AssetResourceName`
 * @example const asset: ResourceNames.AssetResourceName = ResourceNames.asset(10987417, 21974834)
 */
function asset(customerId, assetId) {
    return `customers/${customerId}/assets/${assetId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} fieldType
 * @returns `AssetFieldTypeViewResourceName`
 * @example const assetFieldTypeView: ResourceNames.AssetFieldTypeViewResourceName = ResourceNames.assetFieldTypeView(10987417, 21974834)
 */
function assetFieldTypeView(customerId, fieldType) {
    return `customers/${customerId}/assetFieldTypeViews/${fieldType}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} assetGroupId
 * @returns `AssetGroupResourceName`
 * @example const assetGroup: ResourceNames.AssetGroupResourceName = ResourceNames.assetGroup(10987417, 21974834)
 */
function assetGroup(customerId, assetGroupId) {
    return `customers/${customerId}/assetGroups/${assetGroupId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} assetGroupId
 * @param {string | number} assetId
 * @param {string | number} fieldType
 * @returns `AssetGroupAssetResourceName`
 * @example const assetGroupAsset: ResourceNames.AssetGroupAssetResourceName = ResourceNames.assetGroupAsset(10987417, 21974834, 43949668, 87899336)
 */
function assetGroupAsset(customerId, assetGroupId, assetId, fieldType) {
    return `customers/${customerId}/assetGroupAssets/${assetGroupId}~${assetId}~${fieldType}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} assetGroupId
 * @param {string | number} listingGroupFilterId
 * @returns `AssetGroupListingGroupFilterResourceName`
 * @example const assetGroupListingGroupFilter: ResourceNames.AssetGroupListingGroupFilterResourceName = ResourceNames.assetGroupListingGroupFilter(10987417, 21974834, 43949668)
 */
function assetGroupListingGroupFilter(customerId, assetGroupId, listingGroupFilterId) {
    return `customers/${customerId}/assetGroupListingGroupFilters/${assetGroupId}~${listingGroupFilterId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} assetGroupId
 * @param {string | number} listingGroupFilterId
 * @returns `AssetGroupProductGroupViewResourceName`
 * @example const assetGroupProductGroupView: ResourceNames.AssetGroupProductGroupViewResourceName = ResourceNames.assetGroupProductGroupView(10987417, 21974834, 43949668)
 */
function assetGroupProductGroupView(customerId, assetGroupId, listingGroupFilterId) {
    return `customers/${customerId}/assetGroupProductGroupViews/${assetGroupId}~${listingGroupFilterId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} assetGroupId
 * @param {string | number} criterionId
 * @returns `AssetGroupSignalResourceName`
 * @example const assetGroupSignal: ResourceNames.AssetGroupSignalResourceName = ResourceNames.assetGroupSignal(10987417, 21974834, 43949668)
 */
function assetGroupSignal(customerId, assetGroupId, criterionId) {
    return `customers/${customerId}/assetGroupSignals/${assetGroupId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} assetGroupId
 * @param {string | number} assetCombinationCategory
 * @returns `AssetGroupTopCombinationViewResourceName`
 * @example const assetGroupTopCombinationView: ResourceNames.AssetGroupTopCombinationViewResourceName = ResourceNames.assetGroupTopCombinationView(10987417, 21974834, 43949668)
 */
function assetGroupTopCombinationView(customerId, assetGroupId, assetCombinationCategory) {
    return `customers/${customerId}/assetGroupTopCombinationViews/${assetGroupId}~${assetCombinationCategory}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} assetSetId
 * @returns `AssetSetResourceName`
 * @example const assetSet: ResourceNames.AssetSetResourceName = ResourceNames.assetSet(10987417, 21974834)
 */
function assetSet(customerId, assetSetId) {
    return `customers/${customerId}/assetSets/${assetSetId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} assetSetId
 * @param {string | number} assetId
 * @returns `AssetSetAssetResourceName`
 * @example const assetSetAsset: ResourceNames.AssetSetAssetResourceName = ResourceNames.assetSetAsset(10987417, 21974834, 43949668)
 */
function assetSetAsset(customerId, assetSetId, assetId) {
    return `customers/${customerId}/assetSetAssets/${assetSetId}~${assetId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} assetSetType
 * @returns `AssetSetTypeViewResourceName`
 * @example const assetSetTypeView: ResourceNames.AssetSetTypeViewResourceName = ResourceNames.assetSetTypeView(10987417, 21974834)
 */
function assetSetTypeView(customerId, assetSetType) {
    return `customers/${customerId}/assetSetTypeViews/${assetSetType}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} audienceId
 * @returns `AudienceResourceName`
 * @example const audience: ResourceNames.AudienceResourceName = ResourceNames.audience(10987417, 21974834)
 */
function audience(customerId, audienceId) {
    return `customers/${customerId}/audiences/${audienceId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} batchJobId
 * @returns `BatchJobResourceName`
 * @example const batchJob: ResourceNames.BatchJobResourceName = ResourceNames.batchJob(10987417, 21974834)
 */
function batchJob(customerId, batchJobId) {
    return `customers/${customerId}/batchJobs/${batchJobId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} seasonalityEventId
 * @returns `BiddingDataExclusionResourceName`
 * @example const biddingDataExclusion: ResourceNames.BiddingDataExclusionResourceName = ResourceNames.biddingDataExclusion(10987417, 21974834)
 */
function biddingDataExclusion(customerId, seasonalityEventId) {
    return `customers/${customerId}/biddingDataExclusions/${seasonalityEventId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} seasonalityEventId
 * @returns `BiddingSeasonalityAdjustmentResourceName`
 * @example const biddingSeasonalityAdjustment: ResourceNames.BiddingSeasonalityAdjustmentResourceName = ResourceNames.biddingSeasonalityAdjustment(10987417, 21974834)
 */
function biddingSeasonalityAdjustment(customerId, seasonalityEventId) {
    return `customers/${customerId}/biddingSeasonalityAdjustments/${seasonalityEventId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} biddingStrategyId
 * @returns `BiddingStrategyResourceName`
 * @example const biddingStrategy: ResourceNames.BiddingStrategyResourceName = ResourceNames.biddingStrategy(10987417, 21974834)
 */
function biddingStrategy(customerId, biddingStrategyId) {
    return `customers/${customerId}/biddingStrategies/${biddingStrategyId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} biddingStrategyId
 * @param {string | number} type
 * @param {string | number} modificationMethod
 * @param {string | number} startDate
 * @param {string | number} endDate
 * @returns `BiddingStrategySimulationResourceName`
 * @example const biddingStrategySimulation: ResourceNames.BiddingStrategySimulationResourceName = ResourceNames.biddingStrategySimulation(10987417, 21974834, 43949668, 87899336, 175798672, 351597344)
 */
function biddingStrategySimulation(customerId, biddingStrategyId, type, modificationMethod, startDate, endDate) {
    return `customers/${customerId}/biddingStrategySimulations/${biddingStrategyId}~${type}~${modificationMethod}~${startDate}~${endDate}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} billingSetupId
 * @returns `BillingSetupResourceName`
 * @example const billingSetup: ResourceNames.BillingSetupResourceName = ResourceNames.billingSetup(10987417, 21974834)
 */
function billingSetup(customerId, billingSetupId) {
    return `customers/${customerId}/billingSetups/${billingSetupId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} callDetailId
 * @returns `CallViewResourceName`
 * @example const callView: ResourceNames.CallViewResourceName = ResourceNames.callView(10987417, 21974834)
 */
function callView(customerId, callDetailId) {
    return `customers/${customerId}/callViews/${callDetailId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @returns `CampaignResourceName`
 * @example const campaign: ResourceNames.CampaignResourceName = ResourceNames.campaign(10987417, 21974834)
 */
function campaign(customerId, campaignId) {
    return `customers/${customerId}/campaigns/${campaignId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} assetId
 * @param {string | number} assetLinkSource
 * @param {string | number} fieldType
 * @returns `CampaignAggregateAssetViewResourceName`
 * @example const campaignAggregateAssetView: ResourceNames.CampaignAggregateAssetViewResourceName = ResourceNames.campaignAggregateAssetView(10987417, 21974834, 43949668, 87899336, 175798672)
 */
function campaignAggregateAssetView(customerId, campaignId, assetId, assetLinkSource, fieldType) {
    return `customers/${customerId}/campaignAggregateAssetViews/${campaignId}~${assetId}~${assetLinkSource}~${fieldType}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} assetId
 * @param {string | number} fieldType
 * @returns `CampaignAssetResourceName`
 * @example const campaignAsset: ResourceNames.CampaignAssetResourceName = ResourceNames.campaignAsset(10987417, 21974834, 43949668, 87899336)
 */
function campaignAsset(customerId, campaignId, assetId, fieldType) {
    return `customers/${customerId}/campaignAssets/${campaignId}~${assetId}~${fieldType}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} assetSetId
 * @returns `CampaignAssetSetResourceName`
 * @example const campaignAssetSet: ResourceNames.CampaignAssetSetResourceName = ResourceNames.campaignAssetSet(10987417, 21974834, 43949668)
 */
function campaignAssetSet(customerId, campaignId, assetSetId) {
    return `customers/${customerId}/campaignAssetSets/${campaignId}~${assetSetId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} criterionId
 * @returns `CampaignAudienceViewResourceName`
 * @example const campaignAudienceView: ResourceNames.CampaignAudienceViewResourceName = ResourceNames.campaignAudienceView(10987417, 21974834, 43949668)
 */
function campaignAudienceView(customerId, campaignId, criterionId) {
    return `customers/${customerId}/campaignAudienceViews/${campaignId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} criterionId
 * @returns `CampaignBidModifierResourceName`
 * @example const campaignBidModifier: ResourceNames.CampaignBidModifierResourceName = ResourceNames.campaignBidModifier(10987417, 21974834, 43949668)
 */
function campaignBidModifier(customerId, campaignId, criterionId) {
    return `customers/${customerId}/campaignBidModifiers/${campaignId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignBudgetId
 * @returns `CampaignBudgetResourceName`
 * @example const campaignBudget: ResourceNames.CampaignBudgetResourceName = ResourceNames.campaignBudget(10987417, 21974834)
 */
function campaignBudget(customerId, campaignBudgetId) {
    return `customers/${customerId}/campaignBudgets/${campaignBudgetId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} category
 * @param {string | number} source
 * @returns `CampaignConversionGoalResourceName`
 * @example const campaignConversionGoal: ResourceNames.CampaignConversionGoalResourceName = ResourceNames.campaignConversionGoal(10987417, 21974834, 43949668, 87899336)
 */
function campaignConversionGoal(customerId, campaignId, category, source) {
    return `customers/${customerId}/campaignConversionGoals/${campaignId}~${category}~${source}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} criterionId
 * @returns `CampaignCriterionResourceName`
 * @example const campaignCriterion: ResourceNames.CampaignCriterionResourceName = ResourceNames.campaignCriterion(10987417, 21974834, 43949668)
 */
function campaignCriterion(customerId, campaignId, criterionId) {
    return `customers/${customerId}/campaignCriteria/${campaignId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} customizerAttributeId
 * @returns `CampaignCustomizerResourceName`
 * @example const campaignCustomizer: ResourceNames.CampaignCustomizerResourceName = ResourceNames.campaignCustomizer(10987417, 21974834, 43949668)
 */
function campaignCustomizer(customerId, campaignId, customizerAttributeId) {
    return `customers/${customerId}/campaignCustomizers/${campaignId}~${customizerAttributeId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} baseCampaignId
 * @param {string | number} draftId
 * @returns `CampaignDraftResourceName`
 * @example const campaignDraft: ResourceNames.CampaignDraftResourceName = ResourceNames.campaignDraft(10987417, 21974834, 43949668)
 */
function campaignDraft(customerId, baseCampaignId, draftId) {
    return `customers/${customerId}/campaignDrafts/${baseCampaignId}~${draftId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignGroupId
 * @returns `CampaignGroupResourceName`
 * @example const campaignGroup: ResourceNames.CampaignGroupResourceName = ResourceNames.campaignGroup(10987417, 21974834)
 */
function campaignGroup(customerId, campaignGroupId) {
    return `customers/${customerId}/campaignGroups/${campaignGroupId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} labelId
 * @returns `CampaignLabelResourceName`
 * @example const campaignLabel: ResourceNames.CampaignLabelResourceName = ResourceNames.campaignLabel(10987417, 21974834, 43949668)
 */
function campaignLabel(customerId, campaignId, labelId) {
    return `customers/${customerId}/campaignLabels/${campaignId}~${labelId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @returns `CampaignLifecycleGoalResourceName`
 * @example const campaignLifecycleGoal: ResourceNames.CampaignLifecycleGoalResourceName = ResourceNames.campaignLifecycleGoal(10987417, 21974834)
 */
function campaignLifecycleGoal(customerId, campaignId) {
    return `customers/${customerId}/campaignLifecycleGoals/${campaignId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} clusterId
 * @returns `CampaignSearchTermInsightResourceName`
 * @example const campaignSearchTermInsight: ResourceNames.CampaignSearchTermInsightResourceName = ResourceNames.campaignSearchTermInsight(10987417, 21974834, 43949668)
 */
function campaignSearchTermInsight(customerId, campaignId, clusterId) {
    return `customers/${customerId}/campaignSearchTermInsights/${campaignId}~${clusterId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} query
 * @returns `CampaignSearchTermViewResourceName`
 * @example const campaignSearchTermView: ResourceNames.CampaignSearchTermViewResourceName = ResourceNames.campaignSearchTermView(10987417, 21974834, 43949668)
 */
function campaignSearchTermView(customerId, campaignId, query) {
    return `customers/${customerId}/campaignSearchTermViews/${campaignId}~${query}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} sharedSetId
 * @returns `CampaignSharedSetResourceName`
 * @example const campaignSharedSet: ResourceNames.CampaignSharedSetResourceName = ResourceNames.campaignSharedSet(10987417, 21974834, 43949668)
 */
function campaignSharedSet(customerId, campaignId, sharedSetId) {
    return `customers/${customerId}/campaignSharedSets/${campaignId}~${sharedSetId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} type
 * @param {string | number} modificationMethod
 * @param {string | number} startDate
 * @param {string | number} endDate
 * @returns `CampaignSimulationResourceName`
 * @example const campaignSimulation: ResourceNames.CampaignSimulationResourceName = ResourceNames.campaignSimulation(10987417, 21974834, 43949668, 87899336, 175798672, 351597344)
 */
function campaignSimulation(customerId, campaignId, type, modificationMethod, startDate, endDate) {
    return `customers/${customerId}/campaignSimulations/${campaignId}~${type}~${modificationMethod}~${startDate}~${endDate}`;
}
/**
 * @param {string | number} criterionId
 * @returns `CarrierConstantResourceName`
 * @example const carrierConstant: ResourceNames.CarrierConstantResourceName = ResourceNames.carrierConstant(10987417)
 */
function carrierConstant(criterionId) {
    return `carrierConstants/${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} timestampMicros
 * @param {string | number} commandIndex
 * @param {string | number} mutateIndex
 * @returns `ChangeEventResourceName`
 * @example const changeEvent: ResourceNames.ChangeEventResourceName = ResourceNames.changeEvent(10987417, 21974834, 43949668, 87899336)
 */
function changeEvent(customerId, timestampMicros, commandIndex, mutateIndex) {
    return `customers/${customerId}/changeEvents/${timestampMicros}~${commandIndex}~${mutateIndex}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} changeStatusId
 * @returns `ChangeStatusResourceName`
 * @example const changeStatus: ResourceNames.ChangeStatusResourceName = ResourceNames.changeStatus(10987417, 21974834)
 */
function changeStatus(customerId, changeStatusId) {
    return `customers/${customerId}/changeStatus/${changeStatusId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} advertisingChannelType
 * @param {string | number} assetId
 * @param {string | number} assetSource
 * @param {string | number} fieldType
 * @returns `ChannelAggregateAssetViewResourceName`
 * @example const channelAggregateAssetView: ResourceNames.ChannelAggregateAssetViewResourceName = ResourceNames.channelAggregateAssetView(10987417, 21974834, 43949668, 87899336, 175798672)
 */
function channelAggregateAssetView(customerId, advertisingChannelType, assetId, assetSource, fieldType) {
    return `customers/${customerId}/channelAggregateAssetViews/${advertisingChannelType}~${assetId}~${assetSource}~${fieldType}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} date
 * @param {string | number} gclid
 * @returns `ClickViewResourceName`
 * @example const clickView: ResourceNames.ClickViewResourceName = ResourceNames.clickView(10987417, 21974834, 43949668)
 */
function clickView(customerId, date, gclid) {
    return `customers/${customerId}/clickViews/${date}~${gclid}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} combinedAudienceId
 * @returns `CombinedAudienceResourceName`
 * @example const combinedAudience: ResourceNames.CombinedAudienceResourceName = ResourceNames.combinedAudience(10987417, 21974834)
 */
function combinedAudience(customerId, combinedAudienceId) {
    return `customers/${customerId}/combinedAudiences/${combinedAudienceId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @returns `ContentCriterionViewResourceName`
 * @example const contentCriterionView: ResourceNames.ContentCriterionViewResourceName = ResourceNames.contentCriterionView(10987417, 21974834, 43949668)
 */
function contentCriterionView(customerId, adGroupId, criterionId) {
    return `customers/${customerId}/contentCriterionViews/${adGroupId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} conversionActionId
 * @returns `ConversionActionResourceName`
 * @example const conversionAction: ResourceNames.ConversionActionResourceName = ResourceNames.conversionAction(10987417, 21974834)
 */
function conversionAction(customerId, conversionActionId) {
    return `customers/${customerId}/conversionActions/${conversionActionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} conversionCustomVariableId
 * @returns `ConversionCustomVariableResourceName`
 * @example const conversionCustomVariable: ResourceNames.ConversionCustomVariableResourceName = ResourceNames.conversionCustomVariable(10987417, 21974834)
 */
function conversionCustomVariable(customerId, conversionCustomVariableId) {
    return `customers/${customerId}/conversionCustomVariables/${conversionCustomVariableId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @returns `ConversionGoalCampaignConfigResourceName`
 * @example const conversionGoalCampaignConfig: ResourceNames.ConversionGoalCampaignConfigResourceName = ResourceNames.conversionGoalCampaignConfig(10987417, 21974834)
 */
function conversionGoalCampaignConfig(customerId, campaignId) {
    return `customers/${customerId}/conversionGoalCampaignConfigs/${campaignId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} conversionValueRuleId
 * @returns `ConversionValueRuleResourceName`
 * @example const conversionValueRule: ResourceNames.ConversionValueRuleResourceName = ResourceNames.conversionValueRule(10987417, 21974834)
 */
function conversionValueRule(customerId, conversionValueRuleId) {
    return `customers/${customerId}/conversionValueRules/${conversionValueRuleId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} conversionValueRuleSetId
 * @returns `ConversionValueRuleSetResourceName`
 * @example const conversionValueRuleSet: ResourceNames.ConversionValueRuleSetResourceName = ResourceNames.conversionValueRuleSet(10987417, 21974834)
 */
function conversionValueRuleSet(customerId, conversionValueRuleSetId) {
    return `customers/${customerId}/conversionValueRuleSets/${conversionValueRuleSetId}`;
}
/**
 * @param {string | number} code
 * @returns `CurrencyConstantResourceName`
 * @example const currencyConstant: ResourceNames.CurrencyConstantResourceName = ResourceNames.currencyConstant(10987417)
 */
function currencyConstant(code) {
    return `currencyConstants/${code}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} customAudienceId
 * @returns `CustomAudienceResourceName`
 * @example const customAudience: ResourceNames.CustomAudienceResourceName = ResourceNames.customAudience(10987417, 21974834)
 */
function customAudience(customerId, customAudienceId) {
    return `customers/${customerId}/customAudiences/${customAudienceId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} goalId
 * @returns `CustomConversionGoalResourceName`
 * @example const customConversionGoal: ResourceNames.CustomConversionGoalResourceName = ResourceNames.customConversionGoal(10987417, 21974834)
 */
function customConversionGoal(customerId, goalId) {
    return `customers/${customerId}/customConversionGoals/${goalId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} customInterestId
 * @returns `CustomInterestResourceName`
 * @example const customInterest: ResourceNames.CustomInterestResourceName = ResourceNames.customInterest(10987417, 21974834)
 */
function customInterest(customerId, customInterestId) {
    return `customers/${customerId}/customInterests/${customInterestId}`;
}
/**
 * @param {string | number} customerId
 * @returns `CustomerResourceName`
 * @example const customer: ResourceNames.CustomerResourceName = ResourceNames.customer(10987417)
 */
function customer(customerId) {
    return `customers/${customerId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} assetId
 * @param {string | number} fieldType
 * @returns `CustomerAssetResourceName`
 * @example const customerAsset: ResourceNames.CustomerAssetResourceName = ResourceNames.customerAsset(10987417, 21974834, 43949668)
 */
function customerAsset(customerId, assetId, fieldType) {
    return `customers/${customerId}/customerAssets/${assetId}~${fieldType}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} assetSetId
 * @returns `CustomerAssetSetResourceName`
 * @example const customerAssetSet: ResourceNames.CustomerAssetSetResourceName = ResourceNames.customerAssetSet(10987417, 21974834)
 */
function customerAssetSet(customerId, assetSetId) {
    return `customers/${customerId}/customerAssetSets/${assetSetId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} clientCustomerId
 * @returns `CustomerClientResourceName`
 * @example const customerClient: ResourceNames.CustomerClientResourceName = ResourceNames.customerClient(10987417, 21974834)
 */
function customerClient(customerId, clientCustomerId) {
    return `customers/${customerId}/customerClients/${clientCustomerId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} clientCustomerId
 * @param {string | number} managerLinkId
 * @returns `CustomerClientLinkResourceName`
 * @example const customerClientLink: ResourceNames.CustomerClientLinkResourceName = ResourceNames.customerClientLink(10987417, 21974834, 43949668)
 */
function customerClientLink(customerId, clientCustomerId, managerLinkId) {
    return `customers/${customerId}/customerClientLinks/${clientCustomerId}~${managerLinkId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} category
 * @param {string | number} source
 * @returns `CustomerConversionGoalResourceName`
 * @example const customerConversionGoal: ResourceNames.CustomerConversionGoalResourceName = ResourceNames.customerConversionGoal(10987417, 21974834, 43949668)
 */
function customerConversionGoal(customerId, category, source) {
    return `customers/${customerId}/customerConversionGoals/${category}~${source}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} customizerAttributeId
 * @returns `CustomerCustomizerResourceName`
 * @example const customerCustomizer: ResourceNames.CustomerCustomizerResourceName = ResourceNames.customerCustomizer(10987417, 21974834)
 */
function customerCustomizer(customerId, customizerAttributeId) {
    return `customers/${customerId}/customerCustomizers/${customizerAttributeId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} labelId
 * @returns `CustomerLabelResourceName`
 * @example const customerLabel: ResourceNames.CustomerLabelResourceName = ResourceNames.customerLabel(10987417, 21974834)
 */
function customerLabel(customerId, labelId) {
    return `customers/${customerId}/customerLabels/${labelId}`;
}
/**
 * @param {string | number} customerId
 * @returns `CustomerLifecycleGoalResourceName`
 * @example const customerLifecycleGoal: ResourceNames.CustomerLifecycleGoalResourceName = ResourceNames.customerLifecycleGoal(10987417)
 */
function customerLifecycleGoal(customerId) {
    return `customers/${customerId}/customerLifecycleGoals`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} managerCustomerId
 * @param {string | number} managerLinkId
 * @returns `CustomerManagerLinkResourceName`
 * @example const customerManagerLink: ResourceNames.CustomerManagerLinkResourceName = ResourceNames.customerManagerLink(10987417, 21974834, 43949668)
 */
function customerManagerLink(customerId, managerCustomerId, managerLinkId) {
    return `customers/${customerId}/customerManagerLinks/${managerCustomerId}~${managerLinkId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} criterionId
 * @returns `CustomerNegativeCriterionResourceName`
 * @example const customerNegativeCriterion: ResourceNames.CustomerNegativeCriterionResourceName = ResourceNames.customerNegativeCriterion(10987417, 21974834)
 */
function customerNegativeCriterion(customerId, criterionId) {
    return `customers/${customerId}/customerNegativeCriteria/${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} clusterId
 * @returns `CustomerSearchTermInsightResourceName`
 * @example const customerSearchTermInsight: ResourceNames.CustomerSearchTermInsightResourceName = ResourceNames.customerSearchTermInsight(10987417, 21974834)
 */
function customerSearchTermInsight(customerId, clusterId) {
    return `customers/${customerId}/customerSearchTermInsights/${clusterId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} accountLinkId
 * @returns `CustomerSkAdNetworkConversionValueSchemaResourceName`
 * @example const customerSkAdNetworkConversionValueSchema: ResourceNames.CustomerSkAdNetworkConversionValueSchemaResourceName = ResourceNames.customerSkAdNetworkConversionValueSchema(10987417, 21974834)
 */
function customerSkAdNetworkConversionValueSchema(customerId, accountLinkId) {
    return `customers/${customerId}/customerSkAdNetworkConversionValueSchemas/${accountLinkId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} userId
 * @returns `CustomerUserAccessResourceName`
 * @example const customerUserAccess: ResourceNames.CustomerUserAccessResourceName = ResourceNames.customerUserAccess(10987417, 21974834)
 */
function customerUserAccess(customerId, userId) {
    return `customers/${customerId}/customerUserAccesses/${userId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} invitationId
 * @returns `CustomerUserAccessInvitationResourceName`
 * @example const customerUserAccessInvitation: ResourceNames.CustomerUserAccessInvitationResourceName = ResourceNames.customerUserAccessInvitation(10987417, 21974834)
 */
function customerUserAccessInvitation(customerId, invitationId) {
    return `customers/${customerId}/customerUserAccessInvitations/${invitationId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} customizerAttributeId
 * @returns `CustomizerAttributeResourceName`
 * @example const customizerAttribute: ResourceNames.CustomizerAttributeResourceName = ResourceNames.customizerAttribute(10987417, 21974834)
 */
function customizerAttribute(customerId, customizerAttributeId) {
    return `customers/${customerId}/customizerAttributes/${customizerAttributeId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} productLinkId
 * @param {string | number} dataLinkId
 * @returns `DataLinkResourceName`
 * @example const dataLink: ResourceNames.DataLinkResourceName = ResourceNames.dataLink(10987417, 21974834, 43949668)
 */
function dataLink(customerId, productLinkId, dataLinkId) {
    return `customers/${customerId}/dataLinks/${productLinkId}~${dataLinkId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} placementFingerprint
 * @returns `DetailContentSuitabilityPlacementViewResourceName`
 * @example const detailContentSuitabilityPlacementView: ResourceNames.DetailContentSuitabilityPlacementViewResourceName = ResourceNames.detailContentSuitabilityPlacementView(10987417, 21974834)
 */
function detailContentSuitabilityPlacementView(customerId, placementFingerprint) {
    return `customers/${customerId}/detailContentSuitabilityPlacementViews/${placementFingerprint}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} base64Placement
 * @returns `DetailPlacementViewResourceName`
 * @example const detailPlacementView: ResourceNames.DetailPlacementViewResourceName = ResourceNames.detailPlacementView(10987417, 21974834, 43949668)
 */
function detailPlacementView(customerId, adGroupId, base64Placement) {
    return `customers/${customerId}/detailPlacementViews/${adGroupId}~${base64Placement}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} detailedDemographicId
 * @returns `DetailedDemographicResourceName`
 * @example const detailedDemographic: ResourceNames.DetailedDemographicResourceName = ResourceNames.detailedDemographic(10987417, 21974834)
 */
function detailedDemographic(customerId, detailedDemographicId) {
    return `customers/${customerId}/detailedDemographics/${detailedDemographicId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @returns `DisplayKeywordViewResourceName`
 * @example const displayKeywordView: ResourceNames.DisplayKeywordViewResourceName = ResourceNames.displayKeywordView(10987417, 21974834, 43949668)
 */
function displayKeywordView(customerId, adGroupId, criterionId) {
    return `customers/${customerId}/displayKeywordViews/${adGroupId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} placeholderChainId
 * @param {string | number} distanceBucket
 * @returns `DistanceViewResourceName`
 * @example const distanceView: ResourceNames.DistanceViewResourceName = ResourceNames.distanceView(10987417, 21974834, 43949668)
 */
function distanceView(customerId, placeholderChainId, distanceBucket) {
    return `customers/${customerId}/distanceViews/${placeholderChainId}~${distanceBucket}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} base64Category
 * @param {string | number} languageCode
 * @returns `DomainCategoryResourceName`
 * @example const domainCategory: ResourceNames.DomainCategoryResourceName = ResourceNames.domainCategory(10987417, 21974834, 43949668, 87899336)
 */
function domainCategory(customerId, campaignId, base64Category, languageCode) {
    return `customers/${customerId}/domainCategories/${campaignId}~${base64Category}~${languageCode}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} searchTermFingerprint
 * @param {string | number} headlineFingerprint
 * @param {string | number} landingPageFingerprint
 * @param {string | number} pageUrlFingerprint
 * @returns `DynamicSearchAdsSearchTermViewResourceName`
 * @example const dynamicSearchAdsSearchTermView: ResourceNames.DynamicSearchAdsSearchTermViewResourceName = ResourceNames.dynamicSearchAdsSearchTermView(10987417, 21974834, 43949668, 87899336, 175798672, 351597344)
 */
function dynamicSearchAdsSearchTermView(customerId, adGroupId, searchTermFingerprint, headlineFingerprint, landingPageFingerprint, pageUrlFingerprint) {
    return `customers/${customerId}/dynamicSearchAdsSearchTermViews/${adGroupId}~${searchTermFingerprint}~${headlineFingerprint}~${landingPageFingerprint}~${pageUrlFingerprint}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} expandedFinalUrlFingerprint
 * @returns `ExpandedLandingPageViewResourceName`
 * @example const expandedLandingPageView: ResourceNames.ExpandedLandingPageViewResourceName = ResourceNames.expandedLandingPageView(10987417, 21974834)
 */
function expandedLandingPageView(customerId, expandedFinalUrlFingerprint) {
    return `customers/${customerId}/expandedLandingPageViews/${expandedFinalUrlFingerprint}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} trialId
 * @returns `ExperimentResourceName`
 * @example const experiment: ResourceNames.ExperimentResourceName = ResourceNames.experiment(10987417, 21974834)
 */
function experiment(customerId, trialId) {
    return `customers/${customerId}/experiments/${trialId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} trialId
 * @param {string | number} trialArmId
 * @returns `ExperimentArmResourceName`
 * @example const experimentArm: ResourceNames.ExperimentArmResourceName = ResourceNames.experimentArm(10987417, 21974834, 43949668)
 */
function experimentArm(customerId, trialId, trialArmId) {
    return `customers/${customerId}/experimentArms/${trialId}~${trialArmId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} assetId
 * @param {string | number} fieldType
 * @param {string | number} urlFp
 * @returns `FinalUrlExpansionAssetViewResourceName`
 * @example const finalUrlExpansionAssetView: ResourceNames.FinalUrlExpansionAssetViewResourceName = ResourceNames.finalUrlExpansionAssetView(10987417, 21974834, 43949668, 87899336, 175798672)
 */
function finalUrlExpansionAssetView(customerId, campaignId, assetId, fieldType, urlFp) {
    return `customers/${customerId}/finalUrlExpansionAssetViews/${campaignId}~${assetId}~${fieldType}~${urlFp}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @returns `GenderViewResourceName`
 * @example const genderView: ResourceNames.GenderViewResourceName = ResourceNames.genderView(10987417, 21974834, 43949668)
 */
function genderView(customerId, adGroupId, criterionId) {
    return `customers/${customerId}/genderViews/${adGroupId}~${criterionId}`;
}
/**
 * @param {string | number} criterionId
 * @returns `GeoTargetConstantResourceName`
 * @example const geoTargetConstant: ResourceNames.GeoTargetConstantResourceName = ResourceNames.geoTargetConstant(10987417)
 */
function geoTargetConstant(criterionId) {
    return `geoTargetConstants/${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} countryCriterionId
 * @param {string | number} locationType
 * @returns `GeographicViewResourceName`
 * @example const geographicView: ResourceNames.GeographicViewResourceName = ResourceNames.geographicView(10987417, 21974834, 43949668)
 */
function geographicView(customerId, countryCriterionId, locationType) {
    return `customers/${customerId}/geographicViews/${countryCriterionId}~${locationType}`;
}
/**
 * @param {string | number} googleAdsField
 * @returns `GoogleAdsFieldResourceName`
 * @example const googleAdsField: ResourceNames.GoogleAdsFieldResourceName = ResourceNames.googleAdsField(10987417)
 */
function googleAdsField(googleAdsField) {
    return `googleAdsFields/${googleAdsField}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} placementFingerprint
 * @returns `GroupContentSuitabilityPlacementViewResourceName`
 * @example const groupContentSuitabilityPlacementView: ResourceNames.GroupContentSuitabilityPlacementViewResourceName = ResourceNames.groupContentSuitabilityPlacementView(10987417, 21974834)
 */
function groupContentSuitabilityPlacementView(customerId, placementFingerprint) {
    return `customers/${customerId}/groupContentSuitabilityPlacementViews/${placementFingerprint}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} base64Placement
 * @returns `GroupPlacementViewResourceName`
 * @example const groupPlacementView: ResourceNames.GroupPlacementViewResourceName = ResourceNames.groupPlacementView(10987417, 21974834, 43949668)
 */
function groupPlacementView(customerId, adGroupId, base64Placement) {
    return `customers/${customerId}/groupPlacementViews/${adGroupId}~${base64Placement}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @returns `HotelGroupViewResourceName`
 * @example const hotelGroupView: ResourceNames.HotelGroupViewResourceName = ResourceNames.hotelGroupView(10987417, 21974834, 43949668)
 */
function hotelGroupView(customerId, adGroupId, criterionId) {
    return `customers/${customerId}/hotelGroupViews/${adGroupId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @returns `HotelPerformanceViewResourceName`
 * @example const hotelPerformanceView: ResourceNames.HotelPerformanceViewResourceName = ResourceNames.hotelPerformanceView(10987417)
 */
function hotelPerformanceView(customerId) {
    return `customers/${customerId}/hotelPerformanceView`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} commissionId
 * @returns `HotelReconciliationResourceName`
 * @example const hotelReconciliation: ResourceNames.HotelReconciliationResourceName = ResourceNames.hotelReconciliation(10987417, 21974834)
 */
function hotelReconciliation(customerId, commissionId) {
    return `customers/${customerId}/hotelReconciliations/${commissionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @returns `IncomeRangeViewResourceName`
 * @example const incomeRangeView: ResourceNames.IncomeRangeViewResourceName = ResourceNames.incomeRangeView(10987417, 21974834, 43949668)
 */
function incomeRangeView(customerId, adGroupId, criterionId) {
    return `customers/${customerId}/incomeRangeViews/${adGroupId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} invoiceId
 * @returns `InvoiceResourceName`
 * @example const invoice: ResourceNames.InvoiceResourceName = ResourceNames.invoice(10987417, 21974834)
 */
function invoice(customerId, invoiceId) {
    return `customers/${customerId}/invoices/${invoiceId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} keywordPlanId
 * @returns `KeywordPlanResourceName`
 * @example const keywordPlan: ResourceNames.KeywordPlanResourceName = ResourceNames.keywordPlan(10987417, 21974834)
 */
function keywordPlan(customerId, keywordPlanId) {
    return `customers/${customerId}/keywordPlans/${keywordPlanId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} keywordPlanAdGroupId
 * @returns `KeywordPlanAdGroupResourceName`
 * @example const keywordPlanAdGroup: ResourceNames.KeywordPlanAdGroupResourceName = ResourceNames.keywordPlanAdGroup(10987417, 21974834)
 */
function keywordPlanAdGroup(customerId, keywordPlanAdGroupId) {
    return `customers/${customerId}/keywordPlanAdGroups/${keywordPlanAdGroupId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} keywordPlanAdGroupKeywordId
 * @returns `KeywordPlanAdGroupKeywordResourceName`
 * @example const keywordPlanAdGroupKeyword: ResourceNames.KeywordPlanAdGroupKeywordResourceName = ResourceNames.keywordPlanAdGroupKeyword(10987417, 21974834)
 */
function keywordPlanAdGroupKeyword(customerId, keywordPlanAdGroupKeywordId) {
    return `customers/${customerId}/keywordPlanAdGroupKeywords/${keywordPlanAdGroupKeywordId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} keywordPlanCampaignId
 * @returns `KeywordPlanCampaignResourceName`
 * @example const keywordPlanCampaign: ResourceNames.KeywordPlanCampaignResourceName = ResourceNames.keywordPlanCampaign(10987417, 21974834)
 */
function keywordPlanCampaign(customerId, keywordPlanCampaignId) {
    return `customers/${customerId}/keywordPlanCampaigns/${keywordPlanCampaignId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} keywordPlanCampaignKeywordId
 * @returns `KeywordPlanCampaignKeywordResourceName`
 * @example const keywordPlanCampaignKeyword: ResourceNames.KeywordPlanCampaignKeywordResourceName = ResourceNames.keywordPlanCampaignKeyword(10987417, 21974834)
 */
function keywordPlanCampaignKeyword(customerId, keywordPlanCampaignKeywordId) {
    return `customers/${customerId}/keywordPlanCampaignKeywords/${keywordPlanCampaignKeywordId}`;
}
/**
 * @param {string | number} expressCategoryId
 * @param {string | number} expressSubCategoryId
 * @returns `KeywordThemeConstantResourceName`
 * @example const keywordThemeConstant: ResourceNames.KeywordThemeConstantResourceName = ResourceNames.keywordThemeConstant(10987417, 21974834)
 */
function keywordThemeConstant(expressCategoryId, expressSubCategoryId) {
    return `keywordThemeConstants/${expressCategoryId}~${expressSubCategoryId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @returns `KeywordViewResourceName`
 * @example const keywordView: ResourceNames.KeywordViewResourceName = ResourceNames.keywordView(10987417, 21974834, 43949668)
 */
function keywordView(customerId, adGroupId, criterionId) {
    return `customers/${customerId}/keywordViews/${adGroupId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} labelId
 * @returns `LabelResourceName`
 * @example const label: ResourceNames.LabelResourceName = ResourceNames.label(10987417, 21974834)
 */
function label(customerId, labelId) {
    return `customers/${customerId}/labels/${labelId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} unexpandedFinalUrlFingerprint
 * @returns `LandingPageViewResourceName`
 * @example const landingPageView: ResourceNames.LandingPageViewResourceName = ResourceNames.landingPageView(10987417, 21974834)
 */
function landingPageView(customerId, unexpandedFinalUrlFingerprint) {
    return `customers/${customerId}/landingPageViews/${unexpandedFinalUrlFingerprint}`;
}
/**
 * @param {string | number} criterionId
 * @returns `LanguageConstantResourceName`
 * @example const languageConstant: ResourceNames.LanguageConstantResourceName = ResourceNames.languageConstant(10987417)
 */
function languageConstant(criterionId) {
    return `languageConstants/${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} leadFormUserSubmissionId
 * @returns `LeadFormSubmissionDataResourceName`
 * @example const leadFormSubmissionData: ResourceNames.LeadFormSubmissionDataResourceName = ResourceNames.leadFormSubmissionData(10987417, 21974834)
 */
function leadFormSubmissionData(customerId, leadFormUserSubmissionId) {
    return `customers/${customerId}/leadFormSubmissionData/${leadFormUserSubmissionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} lifeEventId
 * @returns `LifeEventResourceName`
 * @example const lifeEvent: ResourceNames.LifeEventResourceName = ResourceNames.lifeEvent(10987417, 21974834)
 */
function lifeEvent(customerId, lifeEventId) {
    return `customers/${customerId}/lifeEvents/${lifeEventId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} glsEmployeeId
 * @returns `LocalServicesEmployeeResourceName`
 * @example const localServicesEmployee: ResourceNames.LocalServicesEmployeeResourceName = ResourceNames.localServicesEmployee(10987417, 21974834)
 */
function localServicesEmployee(customerId, glsEmployeeId) {
    return `customers/${customerId}/localServicesEmployees/${glsEmployeeId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} localServicesLeadId
 * @returns `LocalServicesLeadResourceName`
 * @example const localServicesLead: ResourceNames.LocalServicesLeadResourceName = ResourceNames.localServicesLead(10987417, 21974834)
 */
function localServicesLead(customerId, localServicesLeadId) {
    return `customers/${customerId}/localServicesLeads/${localServicesLeadId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} localServicesLeadConversationId
 * @returns `LocalServicesLeadConversationResourceName`
 * @example const localServicesLeadConversation: ResourceNames.LocalServicesLeadConversationResourceName = ResourceNames.localServicesLeadConversation(10987417, 21974834)
 */
function localServicesLeadConversation(customerId, localServicesLeadConversationId) {
    return `customers/${customerId}/localServicesLeadConversations/${localServicesLeadConversationId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} glsVerificationArtifactId
 * @returns `LocalServicesVerificationArtifactResourceName`
 * @example const localServicesVerificationArtifact: ResourceNames.LocalServicesVerificationArtifactResourceName = ResourceNames.localServicesVerificationArtifact(10987417, 21974834)
 */
function localServicesVerificationArtifact(customerId, glsVerificationArtifactId) {
    return `customers/${customerId}/localServicesVerificationArtifacts/${glsVerificationArtifactId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @returns `LocationInterestViewResourceName`
 * @example const locationInterestView: ResourceNames.LocationInterestViewResourceName = ResourceNames.locationInterestView(10987417, 21974834, 43949668, 87899336)
 */
function locationInterestView(customerId, campaignId, adGroupId, criterionId) {
    return `customers/${customerId}/locationInterestViews/${campaignId}~${adGroupId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} criterionId
 * @returns `LocationViewResourceName`
 * @example const locationView: ResourceNames.LocationViewResourceName = ResourceNames.locationView(10987417, 21974834, 43949668)
 */
function locationView(customerId, campaignId, criterionId) {
    return `customers/${customerId}/locationViews/${campaignId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @returns `ManagedPlacementViewResourceName`
 * @example const managedPlacementView: ResourceNames.ManagedPlacementViewResourceName = ResourceNames.managedPlacementView(10987417, 21974834, 43949668)
 */
function managedPlacementView(customerId, adGroupId, criterionId) {
    return `customers/${customerId}/managedPlacementViews/${adGroupId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} mediaFileId
 * @returns `MediaFileResourceName`
 * @example const mediaFile: ResourceNames.MediaFileResourceName = ResourceNames.mediaFile(10987417, 21974834)
 */
function mediaFile(customerId, mediaFileId) {
    return `customers/${customerId}/mediaFiles/${mediaFileId}`;
}
/**
 * @param {string | number} mobileAppCategoryId
 * @returns `MobileAppCategoryConstantResourceName`
 * @example const mobileAppCategoryConstant: ResourceNames.MobileAppCategoryConstantResourceName = ResourceNames.mobileAppCategoryConstant(10987417)
 */
function mobileAppCategoryConstant(mobileAppCategoryId) {
    return `mobileAppCategoryConstants/${mobileAppCategoryId}`;
}
/**
 * @param {string | number} criterionId
 * @returns `MobileDeviceConstantResourceName`
 * @example const mobileDeviceConstant: ResourceNames.MobileDeviceConstantResourceName = ResourceNames.mobileDeviceConstant(10987417)
 */
function mobileDeviceConstant(criterionId) {
    return `mobileDeviceConstants/${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} client
 * @returns `OfflineConversionUploadClientSummaryResourceName`
 * @example const offlineConversionUploadClientSummary: ResourceNames.OfflineConversionUploadClientSummaryResourceName = ResourceNames.offlineConversionUploadClientSummary(10987417, 21974834)
 */
function offlineConversionUploadClientSummary(customerId, client) {
    return `customers/${customerId}/offlineConversionUploadClientSummaries/${client}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} conversionTypeId
 * @param {string | number} client
 * @returns `OfflineConversionUploadConversionActionSummaryResourceName`
 * @example const offlineConversionUploadConversionActionSummary: ResourceNames.OfflineConversionUploadConversionActionSummaryResourceName = ResourceNames.offlineConversionUploadConversionActionSummary(10987417, 21974834, 43949668)
 */
function offlineConversionUploadConversionActionSummary(customerId, conversionTypeId, client) {
    return `customers/${customerId}/offlineConversionUploadConversionActionSummaries/${conversionTypeId}~${client}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} offlineUserDataUpdateId
 * @returns `OfflineUserDataJobResourceName`
 * @example const offlineUserDataJob: ResourceNames.OfflineUserDataJobResourceName = ResourceNames.offlineUserDataJob(10987417, 21974834)
 */
function offlineUserDataJob(customerId, offlineUserDataUpdateId) {
    return `customers/${customerId}/offlineUserDataJobs/${offlineUserDataUpdateId}`;
}
/**
 * @param {string | number} criterionId
 * @returns `OperatingSystemVersionConstantResourceName`
 * @example const operatingSystemVersionConstant: ResourceNames.OperatingSystemVersionConstantResourceName = ResourceNames.operatingSystemVersionConstant(10987417)
 */
function operatingSystemVersionConstant(criterionId) {
    return `operatingSystemVersionConstants/${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} adGroupId
 * @param {string | number} base64SearchTerm
 * @returns `PaidOrganicSearchTermViewResourceName`
 * @example const paidOrganicSearchTermView: ResourceNames.PaidOrganicSearchTermViewResourceName = ResourceNames.paidOrganicSearchTermView(10987417, 21974834, 43949668, 87899336)
 */
function paidOrganicSearchTermView(customerId, campaignId, adGroupId, base64SearchTerm) {
    return `customers/${customerId}/paidOrganicSearchTermViews/${campaignId}~${adGroupId}~${base64SearchTerm}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @returns `ParentalStatusViewResourceName`
 * @example const parentalStatusView: ResourceNames.ParentalStatusViewResourceName = ResourceNames.parentalStatusView(10987417, 21974834, 43949668)
 */
function parentalStatusView(customerId, adGroupId, criterionId) {
    return `customers/${customerId}/parentalStatusViews/${adGroupId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} paymentsAccountId
 * @returns `PaymentsAccountResourceName`
 * @example const paymentsAccount: ResourceNames.PaymentsAccountResourceName = ResourceNames.paymentsAccount(10987417, 21974834)
 */
function paymentsAccount(customerId, paymentsAccountId) {
    return `customers/${customerId}/paymentsAccounts/${paymentsAccountId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} placeId
 * @returns `PerStoreViewResourceName`
 * @example const perStoreView: ResourceNames.PerStoreViewResourceName = ResourceNames.perStoreView(10987417, 21974834)
 */
function perStoreView(customerId, placeId) {
    return `customers/${customerId}/perStoreViews/${placeId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} base_64Placement
 * @returns `PerformanceMaxPlacementViewResourceName`
 * @example const performanceMaxPlacementView: ResourceNames.PerformanceMaxPlacementViewResourceName = ResourceNames.performanceMaxPlacementView(10987417, 21974834)
 */
function performanceMaxPlacementView(customerId, base_64Placement) {
    return `customers/${customerId}/performanceMaxPlacementViews/${base_64Placement}`;
}
/**
 * @param {string | number} level
 * @param {string | number} categoryId
 * @returns `ProductCategoryConstantResourceName`
 * @example const productCategoryConstant: ResourceNames.ProductCategoryConstantResourceName = ResourceNames.productCategoryConstant(10987417, 21974834)
 */
function productCategoryConstant(level, categoryId) {
    return `productCategoryConstants/${level}~${categoryId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adgroupId
 * @param {string | number} criterionId
 * @returns `ProductGroupViewResourceName`
 * @example const productGroupView: ResourceNames.ProductGroupViewResourceName = ResourceNames.productGroupView(10987417, 21974834, 43949668)
 */
function productGroupView(customerId, adgroupId, criterionId) {
    return `customers/${customerId}/productGroupViews/${adgroupId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} productLinkId
 * @returns `ProductLinkResourceName`
 * @example const productLink: ResourceNames.ProductLinkResourceName = ResourceNames.productLink(10987417, 21974834)
 */
function productLink(customerId, productLinkId) {
    return `customers/${customerId}/productLinks/${productLinkId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} customerInvitationId
 * @returns `ProductLinkInvitationResourceName`
 * @example const productLinkInvitation: ResourceNames.ProductLinkInvitationResourceName = ResourceNames.productLinkInvitation(10987417, 21974834)
 */
function productLinkInvitation(customerId, customerInvitationId) {
    return `customers/${customerId}/productLinkInvitations/${customerInvitationId}`;
}
/**
 * @param {string | number} qualifyingQuestionId
 * @returns `QualifyingQuestionResourceName`
 * @example const qualifyingQuestion: ResourceNames.QualifyingQuestionResourceName = ResourceNames.qualifyingQuestion(10987417)
 */
function qualifyingQuestion(qualifyingQuestionId) {
    return `qualifyingQuestions/${qualifyingQuestionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} recommendationId
 * @returns `RecommendationResourceName`
 * @example const recommendation: ResourceNames.RecommendationResourceName = ResourceNames.recommendation(10987417, 21974834)
 */
function recommendation(customerId, recommendationId) {
    return `customers/${customerId}/recommendations/${recommendationId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} recommendationType
 * @returns `RecommendationSubscriptionResourceName`
 * @example const recommendationSubscription: ResourceNames.RecommendationSubscriptionResourceName = ResourceNames.recommendationSubscription(10987417, 21974834)
 */
function recommendationSubscription(customerId, recommendationType) {
    return `customers/${customerId}/recommendationSubscriptions/${recommendationType}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} remarketingActionId
 * @returns `RemarketingActionResourceName`
 * @example const remarketingAction: ResourceNames.RemarketingActionResourceName = ResourceNames.remarketingAction(10987417, 21974834)
 */
function remarketingAction(customerId, remarketingActionId) {
    return `customers/${customerId}/remarketingActions/${remarketingActionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} adGroupId
 * @param {string | number} query
 * @returns `SearchTermViewResourceName`
 * @example const searchTermView: ResourceNames.SearchTermViewResourceName = ResourceNames.searchTermView(10987417, 21974834, 43949668, 87899336)
 */
function searchTermView(customerId, campaignId, adGroupId, query) {
    return `customers/${customerId}/searchTermViews/${campaignId}~${adGroupId}~${query}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} sharedSetId
 * @param {string | number} criterionId
 * @returns `SharedCriterionResourceName`
 * @example const sharedCriterion: ResourceNames.SharedCriterionResourceName = ResourceNames.sharedCriterion(10987417, 21974834, 43949668)
 */
function sharedCriterion(customerId, sharedSetId, criterionId) {
    return `customers/${customerId}/sharedCriteria/${sharedSetId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} sharedSetId
 * @returns `SharedSetResourceName`
 * @example const sharedSet: ResourceNames.SharedSetResourceName = ResourceNames.sharedSet(10987417, 21974834)
 */
function sharedSet(customerId, sharedSetId) {
    return `customers/${customerId}/sharedSets/${sharedSetId}`;
}
/**
 * @param {string | number} customerId
 * @returns `ShoppingPerformanceViewResourceName`
 * @example const shoppingPerformanceView: ResourceNames.ShoppingPerformanceViewResourceName = ResourceNames.shoppingPerformanceView(10987417)
 */
function shoppingPerformanceView(customerId) {
    return `customers/${customerId}/shoppingPerformanceView`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} merchantCenterId
 * @param {string | number} channel
 * @param {string | number} languageCode
 * @param {string | number} feedLabel
 * @param {string | number} itemId
 * @returns `ShoppingProductResourceName`
 * @example const shoppingProduct: ResourceNames.ShoppingProductResourceName = ResourceNames.shoppingProduct(10987417, 21974834, 43949668, 87899336, 175798672, 351597344)
 */
function shoppingProduct(customerId, merchantCenterId, channel, languageCode, feedLabel, itemId) {
    return `customers/${customerId}/shoppingProducts/${merchantCenterId}~${channel}~${languageCode}~${feedLabel}~${itemId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @param {string | number} query
 * @returns `SmartCampaignSearchTermViewResourceName`
 * @example const smartCampaignSearchTermView: ResourceNames.SmartCampaignSearchTermViewResourceName = ResourceNames.smartCampaignSearchTermView(10987417, 21974834, 43949668)
 */
function smartCampaignSearchTermView(customerId, campaignId, query) {
    return `customers/${customerId}/smartCampaignSearchTermViews/${campaignId}~${query}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} campaignId
 * @returns `SmartCampaignSettingResourceName`
 * @example const smartCampaignSetting: ResourceNames.SmartCampaignSettingResourceName = ResourceNames.smartCampaignSetting(10987417, 21974834)
 */
function smartCampaignSetting(customerId, campaignId) {
    return `customers/${customerId}/smartCampaignSettings/${campaignId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} customerLinkId
 * @returns `ThirdPartyAppAnalyticsLinkResourceName`
 * @example const thirdPartyAppAnalyticsLink: ResourceNames.ThirdPartyAppAnalyticsLinkResourceName = ResourceNames.thirdPartyAppAnalyticsLink(10987417, 21974834)
 */
function thirdPartyAppAnalyticsLink(customerId, customerLinkId) {
    return `customers/${customerId}/thirdPartyAppAnalyticsLinks/${customerLinkId}`;
}
/**
 * @param {string | number} topicId
 * @returns `TopicConstantResourceName`
 * @example const topicConstant: ResourceNames.TopicConstantResourceName = ResourceNames.topicConstant(10987417)
 */
function topicConstant(topicId) {
    return `topicConstants/${topicId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @returns `TopicViewResourceName`
 * @example const topicView: ResourceNames.TopicViewResourceName = ResourceNames.topicView(10987417, 21974834, 43949668)
 */
function topicView(customerId, adGroupId, criterionId) {
    return `customers/${customerId}/topicViews/${adGroupId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @returns `TravelActivityGroupViewResourceName`
 * @example const travelActivityGroupView: ResourceNames.TravelActivityGroupViewResourceName = ResourceNames.travelActivityGroupView(10987417, 21974834, 43949668)
 */
function travelActivityGroupView(customerId, adGroupId, criterionId) {
    return `customers/${customerId}/travelActivityGroupViews/${adGroupId}~${criterionId}`;
}
/**
 * @param {string | number} customerId
 * @returns `TravelActivityPerformanceViewResourceName`
 * @example const travelActivityPerformanceView: ResourceNames.TravelActivityPerformanceViewResourceName = ResourceNames.travelActivityPerformanceView(10987417)
 */
function travelActivityPerformanceView(customerId) {
    return `customers/${customerId}/travelActivityPerformanceViews`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} userInterestId
 * @returns `UserInterestResourceName`
 * @example const userInterest: ResourceNames.UserInterestResourceName = ResourceNames.userInterest(10987417, 21974834)
 */
function userInterest(customerId, userInterestId) {
    return `customers/${customerId}/userInterests/${userInterestId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} userListId
 * @returns `UserListResourceName`
 * @example const userList: ResourceNames.UserListResourceName = ResourceNames.userList(10987417, 21974834)
 */
function userList(customerId, userListId) {
    return `customers/${customerId}/userLists/${userListId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} userListId
 * @param {string | number} semanticLabel
 * @returns `UserListCustomerTypeResourceName`
 * @example const userListCustomerType: ResourceNames.UserListCustomerTypeResourceName = ResourceNames.userListCustomerType(10987417, 21974834, 43949668)
 */
function userListCustomerType(customerId, userListId, semanticLabel) {
    return `customers/${customerId}/userListCustomerTypes/${userListId}~${semanticLabel}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} countryCriterionId
 * @param {string | number} isTargetingLocation
 * @returns `UserLocationViewResourceName`
 * @example const userLocationView: ResourceNames.UserLocationViewResourceName = ResourceNames.userLocationView(10987417, 21974834, 43949668)
 */
function userLocationView(customerId, countryCriterionId, isTargetingLocation) {
    return `customers/${customerId}/userLocationViews/${countryCriterionId}~${isTargetingLocation}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} videoId
 * @returns `VideoResourceName`
 * @example const video: ResourceNames.VideoResourceName = ResourceNames.video(10987417, 21974834)
 */
function video(customerId, videoId) {
    return `customers/${customerId}/videos/${videoId}`;
}
/**
 * @param {string | number} customerId
 * @param {string | number} adGroupId
 * @param {string | number} criterionId
 * @returns `WebpageViewResourceName`
 * @example const webpageView: ResourceNames.WebpageViewResourceName = ResourceNames.webpageView(10987417, 21974834, 43949668)
 */
function webpageView(customerId, adGroupId, criterionId) {
    return `customers/${customerId}/webpageViews/${adGroupId}~${criterionId}`;
}
