// google-ads-api/build/src/protos/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.longrunning = exports.protobuf = exports.GoogleAdsServiceClient = exports.fieldDataTypes = exports.fields = exports.enums = exports.internalEnums = exports.services = exports.resources = exports.errors = exports.common = exports.VERSION = void 0;
const google_ads_node_1 = require("google-ads-node");
const version_1 = require("../version");
// "as vN" is required to avoid type issues later on
exports.VERSION = version_1.googleAdsVersion;
// -- Expose the googleads generated types --
// The namespace path is shortened for ease of use
exports.common = google_ads_node_1.protos.google.ads.googleads.v21.common;
exports.errors = google_ads_node_1.protos.google.ads.googleads.v21.errors;
exports.resources = google_ads_node_1.protos.google.ads.googleads.v21.resources;
exports.services = google_ads_node_1.protos.google.ads.googleads.v21.services;
// We compile enums ourselves, the Google ones should be used internally only
exports.internalEnums = google_ads_node_1.protos.google.ads.googleads.v21.enums;
var enums_1 = require("./autogen/enums");
Object.defineProperty(exports, "enums", { enumerable: true, get: function () { return enums_1.enums; } });
var fields_1 = require("./autogen/fields");
Object.defineProperty(exports, "fields", { enumerable: true, get: function () { return fields_1.fields; } });
Object.defineProperty(exports, "fieldDataTypes", { enumerable: true, get: function () { return fields_1.fieldDataTypes; } });
// Common service used for report/query methods
var google_ads_node_2 = require("google-ads-node");
Object.defineProperty(exports, "GoogleAdsServiceClient", { enumerable: true, get: function () { return google_ads_node_2.GoogleAdsServiceClient; } });
// -- Expose some google protobuf helpers --
exports.protobuf = google_ads_node_1.protos.google.protobuf;
exports.longrunning = google_ads_node_1.protos.google.longrunning;
